element_calcium_imaging/scan.py

```diff
@@ -613,3 +607,142 @@ def estimate_nd2_scan_duration(nd2_scan_obj):
             pathlib.Path(f).relative_to(root_dir).as_posix() for f in scan_filepaths
         ]
         self.ScanFile.insert([{**key, "file_path": f} for f in scan_files])
+
+
+@schema
+class ScanQualityMetrics(dj.Computed):
+    """Metrics to assess quality of Scan.
+
+    Attributes:
+        ScanInfo.Field (foreign key): Primary key from ScanInfo.Field.
+    """
+
+    definition = """
+    -> ScanInfo.Field
+    """
+
+    class FrameMetrics(dj.Part):
+        """Metrics used to evaluate frames.
+
+        Attributes:
+            ScanInfo.Field (foreign key): Primary key from ScanInfo.Field.
+            channel (int): Channel index.
+            min_intensity (longblob): Minimum value of each frame.
+            mean_intensity (longblob): Mean value at each frame.
+            max_intensity (longblob): Maximum value at each frame.
+            contrast (longblob): Contrast of each frame, difference of 99 and 1 percentiles.
+        """
+
+        definition = """
+        -> master
+        channel: int              # Channel index.
+        ---
+        min_intensity: longblob   # Minimum value of each frame.
+        mean_intensity: longblob  # Mean value at each frame.
+        max_intensity: longblob   # Maximum value at each frame.
+        contrast: longblob        # Contrast of each frame, difference of 99 and 1 percentiles.
+        """
+
+    class QuantalSize(dj.Part):
+        """Quantities inferred from Photon Transfer Curve.
+
+        Attributes:
+            ScanInfo.Field (foreign key): Primary key from ScanInfo.Field.
+            channel (int): Channel index.
+            min_intensity (int): Minimum value in movie.
+            max_intensity (int): Maximum value in movie.
+            quantal_size (float): Quantal size (variance/mean intensity slope), gain.
+            zero_level (int): Level corresponding to zero (computed from variance dependence).
+            quantal_frame (longblob): Average frame expressed in quanta.
+
+        """
+
+        definition = """ 
+        -> master
+        channel         : int      # Channel index
+        ---
+        min_intensity   : int      # Minimum value in movie.
+        max_intensity   : int      # Maximum value in movie.
+        quantal_size    : float    # Quantal size (variance/mean intensity slope), gain.
+        zero_level      : int      # Level corresponding to zero (computed from variance dependence)
+        quantal_frame   : longblob # Average frame expressed in quanta.
+        """
+
+    def make(self, key):
+        acq_software, nchannels = (Scan * ScanInfo & key).fetch1(
+            "acq_software", "nchannels"
+        )
+
+        self.insert1(key)
+
+        for channel in range(nchannels):
+            if acq_software == "ScanImage":
+                import scanreader
+
+                movie = scanreader.read_scan(get_scan_image_files(key))[
```

Comment:
The movie should be read once outside the loop.

Reply:
Yes! The initial solution assumed single channel. Then I added the loop over channels.

I'm fixing it now.